### PR TITLE
cmd/evm: add difficulty calculation to t8n tool

### LINKED ISSUE #23249: evm t8n  calculate difficulty
The current env format is 
```
{
   "currentCoinbase" : "0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba",
   "currentDifficulty" : "0x020000",
   "currentGasLimit" : "0x05f5e100",
   "currentNumber" : "0x01",
   "currentTimestamp" : "0x03e8",
   "previousHash" : "0xe729de3fec21e30bea3d56adb01ed14bc107273c2775f9355afb10f594a10d9e",
   "blockHashes" : {
       "0" : "0xe729de3fec21e30bea3d56adb01ed14bc107273c2775f9355afb10f594a10d9e"
   }
}
```

want to extend the following format to be acceptable (as well as the old one)
```
{
   "currentCoinbase" : "0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba",
   "parentDifficulty" : "0x020000",
   "currentGasLimit" : "0x05f5e100",
   "currentNumber" : "0x01",
   "parentTimestamp" : "0x03e7",
   "currentTimestamp" : "0x03e8",
   "previousHash" : "0xe729de3fec21e30bea3d56adb01ed14bc107273c2775f9355afb10f594a10d9e",
   "blockHashes" : {
       "0" : "0xe729de3fec21e30bea3d56adb01ed14bc107273c2775f9355afb10f594a10d9e"
   }
}
```

new fields: *parentDifficulty* instead of currentDifficulty
and *parentTimestamp*

the result file then
```
{
  "currentDifficulty" : "0x020001"
  "stateRoot": "0x1c99b01120e7a2fa1301b3505f20100e72362e5ac3f96854420e56ba8984d716",
  "txRoot": "0xb5eee60b45801179cbde3781b9a5dee9b3111554618c9cda3d6f7e351fd41e0b",
  "receiptRoot": "0x86ceb80cb6bef8fe4ac0f1c99409f67cb2554c4432f374e399b94884eb3e6562",
  "logsHash": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "receipts": [
  ]
}
```

would report the new difficulty calculated with geth formula  *currentDifficulty*

**Rationale**
To avoid point of failure when generating the tests. Currently retesteth calculates the difficulty, but retesteth difficulty formula might be wrong. Since we test the clients, it would make sense that retesteth asks t8ntool what geth evm thinks would be the new difficulty given parentTimestamp and parentDifficulty and currentTimestamp.

### PR DESCRIPTION
**Pull request by holiman**
This PR adds functionality to the `evm t8n` to calculate _ethash_ difficulty. If the caller does not provide a `currentDifficulty`, but instead provides the `parentTimestamp` (well, semi-optional, will default to `0` if not given), and `parentDifficulty`, we can calculate it for him. 

The caller can also provide a `parentUncleHash`. In most, but not all cases, the parent uncle hash also affects the formula. If no such hash is provided (or, if the empty all-zero hash is provided), it's assumed that there were no uncles. 

Documented in `testdata/14/readme.md`.

Closes #23249

## Example run

```
[user@work evm]$ ./evm t8n --input.alloc=./testdata/14/alloc.json --input.txs=./testdata/14/txs.json --input.env=./testdata/14/env.json --output.result=stdout --state.fork=London
INFO [08-08|17:35:46.876] Trie dumping started                     root=6f0588..7f4bdc
INFO [08-08|17:35:46.876] Trie dumping complete                    accounts=2 elapsed="89.313µs"
INFO [08-08|17:35:46.877] Wrote file                               file=alloc.json
{
 "result": {
  "stateRoot": "0x6f058887ca01549716789c380ede95aecc510e6d1fdc4dbf67d053c7c07f4bdc",
  "txRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "receiptRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "logsHash": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "receipts": [],
  "currentDifficulty": 3311729559732224
 }
}
```

### CODE REVIEW

**Comment by holiman** [comment]
@winsvega does this work for you?

File: cmd/evm/internal/t8ntool/execution.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ package t8ntool
 
 import (
 	"fmt"
+	"github.com/ethereum/go-ethereum/consensus/ethash"
```
Comment (rjl493456442) [inline cmd/evm/internal/t8ntool/execution.go:21]: please gofmt it

**Comment by holiman** [comment]
@winsvega ping -- does this fit your needs?

**Comment by winsvega** [comment]
I guess. I can chek tomorrow.

**Comment by winsvega** [comment]
we used to trat every field as string. 
the currentDifficulty is returned as int in json

**Comment by holiman** [comment]
@winsvega fixed: 
```
]$  ./evm t8n --input.alloc=./testdata/14/alloc.json --input.txs=./testdata/14/txs.json --input.env=./testdata/14/env.json --output.result=stdout --state.fork=London
INFO [08-25|11:30:17.118] Trie dumping started                     root=6f0588..7f4bdc
INFO [08-25|11:30:17.118] Trie dumping complete                    accounts=2 elapsed="85.624µs"
INFO [08-25|11:30:17.118] Wrote file                               file=alloc.json
{
 "result": {
  "stateRoot": "0x6f058887ca01549716789c380ede95aecc510e6d1fdc4dbf67d053c7c07f4bdc",
  "txRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "receiptRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "logsHash": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "receipts": [],
  "currentDifficulty": "0xbc40020000000"
 }
}
```

File: cmd/evm/internal/t8ntool/transition.go [hunk 2]
```diff
@@ -252,6 +252,19 @@ func Main(ctx *cli.Context) error {
 			return NewError(ErrorVMConfig, errors.New("EIP-1559 config but missing 'currentBaseFee' in env section"))
 		}
 	}
+	if env := prestate.Env; env.Difficulty == nil {
+		// If difficulty was not provided by caller, we need to calculate it.
+		switch {
+		case env.ParentDifficulty == nil:
```
Comment (MariusVanDerWijden) [inline cmd/evm/internal/t8ntool/transition.go:258]: What if we provide the parentDifficulty but not the parentTimestamp

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM, just a tiny question

File: cmd/evm/internal/t8ntool/transition.go (hunk 2 above)
Comment (holiman) [reply cmd/evm/internal/t8ntool/transition.go:258]: parentTimeStamp is a `uint64`, if it's not explicitly provided, it's `0`.

Comment (holiman) [reply cmd/evm/internal/t8ntool/transition.go:258]: But maybe we should have a check that parentTimestamp is older than current...

Comment (holiman) [reply cmd/evm/internal/t8ntool/transition.go:258]: done

**Pull request merged**